Code/Mantid/Framework/MDAlgorithms/src/IntegrateEllipsoids.cpp

```diff
@@ -41,18 +41,28 @@ const std::size_t DIMS(3);
  * @param integrator : itegrator object on which qlists are accumulated
  * @param prog : progress object
  * @param wksp : input EventWorkspace
- * @param unitConverter : Unit converter
- * @param qConverter : Q converter
+ * @param UBinv : inverse of UB matrix
+ * @param hkl_integ ; boolean for integrating in HKL space
  */
-void qListFromEventWS(Integrate3DEvents &integrator, Progress &prog,
+void IntegrateEllipsoids::qListFromEventWS(Integrate3DEvents &integrator, Progress &prog,
                       EventWorkspace_sptr &wksp,
-                      UnitsConversionHelper &unitConverter,
-                      MDTransf_sptr &qConverter, DblMatrix const &UBinv, bool hkl_integ) {
+                      DblMatrix const &UBinv, bool hkl_integ) {
   // loop through the eventlists
-  std::vector<double> buffer(DIMS);
 
-  size_t numSpectra = wksp->getNumberHistograms();
-  for (std::size_t i = 0; i < numSpectra; ++i) {
+  int numSpectra = static_cast<int>(wksp->getNumberHistograms());
+  PARALLEL_FOR1(wksp)
+  for (int i = 0; i < numSpectra; ++i) {
+    PARALLEL_START_INTERUPT_REGION
+
+    // units conversion helper
+    UnitsConversionHelper unitConverter;
+    unitConverter.initialize(m_targWSDescr, "Momentum");
+
+    // initialize the MD coordinates conversion class
+    MDTransfQ3D qConverter;
+    qConverter.initialize(m_targWSDescr);
```

Comment:
This is what caused us to start getting the correct answers.  
From Alex:
but its auxiliary -- *m_pL2s is the pointer to the array of Li, allocated in PreprocessedDetector's table
in OMP you are expected to instanciate as many helpers as there are threads --
with this approach it will work fine
I am doing it like ConvertUnits is doing itself.
the constructor for UnitsConversion has to be within OMP loop -- like ConvertUints class itself.